core/templates/pages/exploration-player-page/services/conversation-flow.service.ts

```diff
@@ -0,0 +1,149 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to manage the conversation flow of the exploration player,
+ * controlling behaviours such as adding cards to the stack and displaying them, or
+ * submitting the answer to progress further.
+ */
+
+import {StateCard} from 'domain/state_card/state-card.model';
+import {Injectable} from '@angular/core';
+import {downgradeInjectable} from '@angular/upgrade/static';
+import {ContentTranslationLanguageService} from './content-translation-language.service';
+import {ContentTranslationManagerService} from './content-translation-manager.service';
+import {ExplorationPlayerStateService} from './exploration-player-state.service';
+import {PlayerPositionService} from './player-position.service';
+import {PlayerTranscriptService} from './player-transcript.service';
+import {ExplorationPlayerConstants} from '../exploration-player-page.constants';
+import {WindowDimensionsService} from 'services/contextual/window-dimensions.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ConversationFlowService {
+  TIME_PADDING_MSEC = 250;
+  TIME_FADEIN_MSEC = 100;
+  TIME_NUM_CARDS_CHANGE_MSEC = 500;
+
+  playerIsAnimatingToTwoCards!: boolean;
+  playerIsAnimatingToOneCard!: boolean;
+
+  constructor(
+    private contentTranslationLanguageService: ContentTranslationLanguageService,
+    private contentTranslationManagerService: ContentTranslationManagerService,
+    private explorationPlayerStateService: ExplorationPlayerStateService,
+    private playerPositionService: PlayerPositionService,
+    private playerTranscriptService: PlayerTranscriptService,
+    private windowDimensionsService: WindowDimensionsService
+  ) {}
+
+  addAndDisplayNewCard(newCard: StateCard): void {
+    this.playerTranscriptService.addNewCard(newCard);
+    const explorationLanguageCode =
+      this.explorationPlayerStateService.getLanguageCode();
+    const selectedLanguageCode =
+      this.contentTranslationLanguageService.getCurrentContentLanguageCode();
+    if (explorationLanguageCode !== selectedLanguageCode) {
+      this.contentTranslationManagerService.displayTranslations(
+        selectedLanguageCode
+      );
+    }
+
+    let totalNumCards = this.playerTranscriptService.getNumCards();
+
+    let previousSupplementalCardIsNonempty =
+      totalNumCards > 1 &&
+      this.isSupplementalCardNonempty(
+        this.playerTranscriptService.getCard(totalNumCards - 2)
+      );
+
+    let nextSupplementalCardIsNonempty = this.isSupplementalCardNonempty(
+      this.playerTranscriptService.getLastCard()
+    );
+
+    // Check if next card is supplemental and previous card isn't, indicating
+    // we need to animate player to two cards if the window can show two cards.
+    if (
+      totalNumCards > 1 &&
+      this.canWindowShowTwoCards() &&
```

Comment:
Remember that the service should be common to all the skins. Not all skins have "one card" and "two cards" though. In fact in the new skin I'm not even sure there's a concept of "supplemental interaction".

So you will probably want to have the logic for "one or two cards" and "supplemental interaction or not" in the conversation skin component, not this common service. This probably involves separating the "add a new card" logic from the "animate to the next thing" logic. One way you can do this is for the controller to pass in a callback that the service passes the next card's info to (so that the controller can determine whether it's a one- or two-card thing and react accordingly).

Similarly, anything about window width should be in the conversation skin layer.

Reply:
Makes sense, but could you kindly clarify a bit on what you mean by all skins not having one card or two cards?

Also regarding supplemental interactions - I don't see them being handled in the Figma file, could you please confirm with the designer if they'll still be a thing, and how the redesign will look like?

Reply:
@seanlip If we choose to have the conditions in the controllers, doesn't that entail that almost the entire function (from lines 64 to 99) including the animation functions are supposed to be duplicated in the new skin's controller? Just confirming to be sure, thanks.

Reply:
I think the controller could have logic like "if(isNextInteractionSupplemental() && !isPreviousInteractionSupplemental()) { ...}" (this is just an example).

So the animation logic and stuff like canWindowShowTwoCards() would stay in the controller but you can get the info about whether the previous/next card is supplemental or not from the service. (Since the service presumably maintains a transcript of cards being shown and the interaction type of each of them.)

Does that help?

Reply:
Done, thanks!